pkg/controller/certificates/trigger/policies/gatherer.go

```diff
@@ -31,40 +31,46 @@ import (
 	"github.com/jetstack/cert-manager/pkg/util/predicate"
 )
 
-// Gatherer is used to gather data about a Certificate in order to evaluate
-// its current readiness/state by applying policy functions to it.
-type Gatherer struct {
-	CertificateRequestLister cmlisters.CertificateRequestLister
-	SecretLister             corelisters.SecretLister
-}
-
-func (g *Gatherer) DataForCertificate(ctx context.Context, crt *cmapi.Certificate) (Input, error) {
+// DataForCertificate is used to gather data about a Certificate in order
+// to evaluate its current readiness/state by applying policy functions to
+// it.
+//
+// The returned input.CurrentRevisionRequest and input.Secret are left nil
+// when they cannot be found. The input.Certificate is copied as-is from
+// the given crt.
+func DataForCertificate(ctx context.Context, getSecret func(string) (*v1.Secret, error), lister cmlisters.CertificateRequestNamespaceLister, crt *cmapi.Certificate) (Input, error) {
 	log := logf.FromContext(ctx)
 	// Attempt to fetch the Secret being managed but tolerate NotFound errors.
-	secret, err := g.SecretLister.Secrets(crt.Namespace).Get(crt.Spec.SecretName)
+	secret, err := getSecret(crt.Spec.SecretName)
```

Comment:
What's the reasoning for getting rid of the `Gatherer` struct here? It seems to make things less obvious as to what they are actually doing, and we can create fake listers/getters that are populated with resources used during test fixtures (similar to other packages in the project)

Reply:
I thought using a flat function (as opposed to a struct `Gatherer` + receiver `DataForCertificate`) would make things a bit simpler. Truth is: it doesn't make things simpler 😁 The list of arguments passed to DataForCertificate became quite long/unreadable

> we can create fake listers/getters that are populated with resources used during test fixtures

You are right; @joshuastern told me the same thing yesterday 😅 

I'll re-write all that and try to
- keep the `Gatherer` struct since it makes `DataForCertificate` itself more readable (less arguments need to be passed = lower visual overhead)
- use a fake clientset in order to test `DataForCertificate` like we do in all the controller tests